dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/report/MessageFormatter.java

```diff
@@ -77,36 +103,36 @@ else if ( MetadataIdentifier.class.isAssignableFrom( argument.getClass() ) )
         }
         else if ( CategoryOptionCombo.class.isAssignableFrom( argument.getClass() ) )
         {
-            return getIdAndName( idSchemeParams.toMetadataIdentifier( (CategoryOptionCombo) argument ),
+            return getIdAndName( idSchemes.toMetadataIdentifier( (CategoryOptionCombo) argument ),
                 (CategoryOptionCombo) argument );
         }
         else if ( CategoryOption.class.isAssignableFrom( argument.getClass() ) )
         {
-            return getIdAndName( idSchemeParams.toMetadataIdentifier( (CategoryOption) argument ),
+            return getIdAndName( idSchemes.toMetadataIdentifier( (CategoryOption) argument ),
                 (CategoryOption) argument );
         }
         else if ( DataElement.class.isAssignableFrom( argument.getClass() ) )
         {
-            return getIdAndName( idSchemeParams.toMetadataIdentifier( (DataElement) argument ),
+            return getIdAndName( idSchemes.toMetadataIdentifier( (DataElement) argument ),
                 (DataElement) argument );
         }
         else if ( OrganisationUnit.class.isAssignableFrom( argument.getClass() ) )
         {
-            return getIdAndName( idSchemeParams.toMetadataIdentifier( (OrganisationUnit) argument ),
+            return getIdAndName( idSchemes.toMetadataIdentifier( (OrganisationUnit) argument ),
                 (OrganisationUnit) argument );
         }
         else if ( Program.class.isAssignableFrom( argument.getClass() ) )
         {
-            return getIdAndName( idSchemeParams.toMetadataIdentifier( (Program) argument ), (Program) argument );
+            return getIdAndName( idSchemes.toMetadataIdentifier( (Program) argument ), (Program) argument );
         }
         else if ( ProgramStage.class.isAssignableFrom( argument.getClass() ) )
         {
-            return getIdAndName( idSchemeParams.toMetadataIdentifier( (ProgramStage) argument ),
+            return getIdAndName( idSchemes.toMetadataIdentifier( (ProgramStage) argument ),
                 (ProgramStage) argument );
         }
         else if ( IdentifiableObject.class.isAssignableFrom( argument.getClass() ) )
         {
-            return getIdAndName( idSchemeParams.toMetadataIdentifier( (IdentifiableObject) argument ),
+            return getIdAndName( idSchemes.toMetadataIdentifier( (IdentifiableObject) argument ),
```

Comment:
All the previous Metadata classes should implement IdentifiableObject.
Would be safe to just remove them?

Reply:
We support overriding the general `idScheme` param with metadata specific one like `dataElementIdScheme`, ... see https://docs.dhis2.org/en/develop/using-the-api/dhis-core-version-master/new-tracker.html#request-parameters

Since generic type information is not available at runtime I used method overloading.
idSchemes.toMetadataIdentifier( IdentifiableObject metadata ) will use the `idScheme` parameter.

idSchemes.toMetadataIdentifier( DataElement metadata ) will use the `dataElementIdScheme` and so on.

So I cannot remove them, otherwise it would not report in the correct idScheme.